### ISSUE TITLE
flat() can hang given small context.bits

### ISSUE THREAD

**Issue by justinsteven** [Gallopsled/pwntools#2184]
Given a small enough `pwn.context.bits`, `pwn.flat({n: m})` will hang for a large enough (but still quite small) `n`

For example, given `context.bits == 32`, `pwn.flat({256: 0x1337})` is fine:

```
>>> with pwn.context.local(bits=32):
...   pwn.flat({256: 0x1337})
...
b'aaaabaaacaaadaaaeaaafaaagaaahaaaiaaajaaakaaalaaamaaanaaaoaaapaaaqaaaraaasaaataaauaaavaaawaaaxaaayaaazaabbaabcaabdaabeaabfaabgaabhaabiaabjaabkaablaabmaabnaaboaabpaabqaabraabsaabtaabuaabvaabwaabxaabyaabzaacbaaccaacdaaceaacfaacgaachaaciaacjaackaaclaacmaacnaac7\x13\x00\x00'
```

The same hangs with `context.bits == 16`:

```
>>> with pwn.context.local(bits=16):
...   pwn.flat({256: 0x1337})
...
[ ... hangs ...]
```

But it's fine with an `n` of 255:

```
>>> with pwn.context.local(bits=16):
...   pwn.flat({255: 0x1337})
...
b'aaaabaaacaaadaaaeaaafaaagaaahaaaiaaajaaakaaalaaamaaanaaaoaaapaaaqaaaraaasaaataaauaaavaaawaaaxaaayaaazaabbaabcaabdaabeaabfaabgaabhaabiaabjaabkaablaabmaabnaaboaabpaabqaabraabsaabtaabuaabvaabwaabxaabyaabzaacbaaccaacdaaceaacfaacgaachaaciaacjaackaaclaacmaacnaa7\x13'
```

This affects architectures such as msp430:

```
>>> with pwn.context.local(arch="msp430"):
...   pwn.flat({256: 0x1337})
...
[... hangs ...]
```

As a more extreme example, the following is ok:

```
>>> with pwn.context.local(bits=8):
...   pwn.flat({0: 0x41})
...
b'A'
```

The following isn't:

```
>>> with pwn.context.local(bits=8):
...   pwn.flat({1: 0x41})
...
[ ... hangs ...]
```

**Comment by Arusekk** [comment]
Can you post a traceback (stack trace) of interrputed program (Ctrl+C)? It should be helpful for initial diagnosis.

Email z czwartku 27. kwietnia 2023 od Justina Stevena:
> Given a small enough `pwn.context.bits`, `pwn.flat({n: m})` will hang for a large enough (but still quite small) `n`
> 
> For example, given `context.bits == 32`, `pwn.flat({256: 0x1337})` is fine:
> 
> ```
> >>> with pwn.context.local(bits=32):
> ...   pwn.flat({256: 0x1337})
> ...
> b'aaaabaaacaaadaaaeaaafaaagaaahaaaiaaajaaakaaalaaamaaanaaaoaaapaaaqaaaraaasaaataaauaaavaaawaaaxaaayaaazaabbaabcaabdaabeaabfaabgaabhaabiaabjaabkaablaabmaabnaaboaabpaabqaabraabsaabtaabuaabvaabwaabxaabyaabzaacbaaccaacdaaceaacfaacgaachaaciaacjaackaaclaacmaacnaac7\x13\x00\x00'
> ```
> 
> The same hangs with `context.bits == 16`:
> 
> ```
> >>> with pwn.context.local(bits=16):
> ...   pwn.flat({256: 0x1337})
> ...
> [ ... hangs ...]
> ```
> 
> But it's fine with an `n` of 255:
> 
> ```
> >>> with pwn.context.local(bits=16):
> ...   pwn.flat({255: 0x1337})
> ...
> b'aaaabaaacaaadaaaeaaafaaagaaahaaaiaaajaaakaaalaaamaaanaaaoaaapaaaqaaaraaasaaataaauaaavaaawaaaxaaayaaazaabbaabcaabdaabeaabfaabgaabhaabiaabjaabkaablaabmaabnaaboaabpaabqaabraabsaabtaabuaabvaabwaabxaabyaabzaacbaaccaacdaaceaacfaacgaachaaciaacjaackaaclaacmaacnaa7\x13'
> ```
> 
> This affects architectures such as msp430:
> 
> ```
> >>> with pwn.context.local(arch="msp430"):
> ...   pwn.flat({256: 0x1337})
> ...
> [... hangs ...]
> ```
> 
> As a more extreme example, the following is ok:
> 
> ```
> >>> with pwn.context.local(bits=8):
> ...   pwn.flat({0: 0x41})
> ...
> b'A'
> ```
> 
> The following isn't:
> 
> ```
> >>> with pwn.context.local(bits=8):
> ...   pwn.flat({1: 0x41})
> ...
> [ ... hangs ...]
> ```
> 
> -- 
> Reply to this email directly or view it on GitHub:
> https://github.com/Gallopsled/pwntools/issues/2184
> You are receiving this because you are subscribed to this thread.
> 
> Message ID: ***@***.***

-- 
Wysłane z mojego urządzenia Sailfish

**Comment by justinsteven** [comment]
```
>>> import pwn

>>> pwn.version
'4.11.0dev'

>>> with pwn.context.local(bits=16):
...   pwn.flat({256: 0x1337})
...
^CTraceback (most recent call last):
  File "<stdin>", line 2, in <module>
  File "/home/justin/pwntools/pwnlib/context/__init__.py", line 1600, in setter
    return function(*a, **kw)
  File "/home/justin/pwntools/pwnlib/util/packing.py", line 775, in flat
    out = _flat(args, preprocessor, make_packer(), filler, stacklevel + 2)
  File "/home/justin/pwntools/pwnlib/util/packing.py", line 590, in _flat
    filler, val = _fit(arg, preprocessor, packer, filler, stacklevel + 1)
  File "/home/justin/pwntools/pwnlib/util/packing.py", line 522, in _fit
    k = fill(pack(k))
  File "/home/justin/pwntools/pwnlib/util/packing.py", line 510, in fill
    pad.append(next(filler))
KeyboardInterrupt

>>> with pwn.context.local(bits=8):
...   pwn.flat({1: 0x41})
...
^CTraceback (most recent call last):
  File "<stdin>", line 2, in <module>
  File "/home/justin/pwntools/pwnlib/context/__init__.py", line 1600, in setter
    return function(*a, **kw)
  File "/home/justin/pwntools/pwnlib/util/packing.py", line 775, in flat
    out = _flat(args, preprocessor, make_packer(), filler, stacklevel + 2)
  File "/home/justin/pwntools/pwnlib/util/packing.py", line 590, in _flat
    filler, val = _fit(arg, preprocessor, packer, filler, stacklevel + 1)
  File "/home/justin/pwntools/pwnlib/util/packing.py", line 522, in _fit
    k = fill(pack(k))
  File "/home/justin/pwntools/pwnlib/util/packing.py", line 510, in fill
    pad.append(next(filler))
KeyboardInterrupt
```

**Comment by Arusekk** [comment]
Would you mind trying to find the cause and submitting a PR with a fix?

**Comment by justinsteven** [comment]
Ah perhaps this is working as designed :(

tl;dr `flat()` says it takes integer dictionary keys that are >= `2**(word_size-8)` and packs them before searching in the De Bruijin sequence. For small `context.bits` values, `flat()` will very readily think that an integer key is supposed to represent part of a cyclic pattern and will try to `cyclic_find()` it? And will keep on looking forever, hence the hang.

And so I don't know what can be done about this. But it's a pretty gnarly edge case.

---

The root cause seems to be near to https://github.com/Gallopsled/pwntools/blob/c1ed0e6f101dceb1fdc0cd78c25a905e4a4e4d5c/pwnlib/util/packing.py#L518-L522

Given the test cases:

```
#!/usr/bin/env python3
import pwn

def main():
    with pwn.context.local(bits=32):
        print(pwn.flat({256: 0x1337}))

    with pwn.context.local(bits=16):
        print(pwn.flat({255: 0x1337}))

    with pwn.context.local(bits=16):
        print(pwn.flat({256: 0x1337}))


if __name__ == "__main__":
    main()
```

For the first case, `large_key = 2**(context.word_size-8) = 2**(32-8) = 16777216` which is greater than 256 and so the branch is not taken.

For the second, `large_key = 2**(context.word_size-8) = 2**(16-8) = 256` which is greater than 255 and so again the branch is not taken.

For the third, because 256 is equal to `large_key`, the branch is taken, we do `k = fill(pack(k))` which ends up being `k = fill(b\x00\x01)`. The De Bruijin sequence will never contain `b'\x00\x01'` and so `fill()` loops forever, consuming CPU and filling memory.

This seems to be to support the following advertised functionality of `flat()` (emphasis mine)

> Dictionary keys give offsets at which to place the corresponding values (which are recursively flattened). Offsets are relative to where the flattened dictionary occurs in the output (i.e. `{0: 'foo'}` is equivalent to `'foo'`). Offsets can be integers, unicode strings or regular strings. **Integer offsets >= `2**(word_size-8)` are converted to a string using `pack()`**. Unicode strings are UTF-8 encoded. After these conversions offsets are either integers or strings. In the latter case, **the offset will be the lowest index at which the string occurs in filler**.

**Comment by Arusekk** [comment]
Alright, I don't know if this can be worked around, but let's keep it the way it is.
